File: src/leetcode/MergeKSortedList.java
package leetcode;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.PriorityQueue;

public class MergeKSortedList {
	public ListNode mergeKLists(ArrayList<ListNode> lists) {
		if(lists == null || lists.size() == 0)
            return null;
		PriorityQueue<ListNode> heap = new PriorityQueue<ListNode>(lists.size(), new Comparator<ListNode>(){
			public int compare(ListNode lhs, ListNode rhs) {
		        if (lhs.val > rhs.val) return +1;
		        if (lhs.val == rhs.val) return 0;
		        return -1;
		    }
		});
		
		ListNode node = null, head = null;
		for(ListNode list : lists){
			if(list!= null)
				heap.add(list);
		}
		
		while(heap.size() > 0){
			ListNode minList = heap.poll();
			if(head == null){
				head = minList;
				node = minList;
			}else{
				node.next = minList;
				node = node.next;
			}
			
			minList = minList.next;
			if(minList!= null)
				heap.add(minList);
 		}
		
		return head;
     }
}

File: python/leetcode/MergeKSortedList.py
class MergeKSortedList:
    def mergeKLists(self, lists):
        heap = []
        for list in lists:
            if list:
                heapq.heappush(heap, ( list.val, list ))
        head = None
        node = None
        while heap:
            val, minNode = heapq.heappop(heap)
            if head is None:
                head = minNode
            else:
                node.next = minNode
            node = minNode
            
            minNode = minNode.next
            if minNode:
                heapq.heappush(heap, (minNode.val, minNode ))
        return head